commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -26,56 +26,55 @@ object JsonStringInput {
-    reader.skipWs()
-    startIdx = reader.index
-    val res = reader.parseValue()
-    res match {
-      case JsonStringInput.ListMarker | JsonStringInput.ObjectMarker =>
-      case _ => afterElement()
-    }
-    res
+  reader.currentValue match {
+    case JsonStringInput.ListMarker | JsonStringInput.ObjectMarker =>
+    case _ => afterElement()
   }
 
-  def jsonType: String = value match {
+  def jsonType: String = reader.currentValue match {
     case JsonStringInput.ListMarker => "list"
     case JsonStringInput.ObjectMarker => "object"
-    case _: String => "string"
+    case _: String => if (reader.isNumber) "number" else "string"
     case _: Boolean => "boolean"
-    case _: Int => "integer number"
-    case _: Double => "decimal number"
     case null => "null"
   }
 
-  private def expected(what: String) = throw new ReadFailure(s"Expected $what but got $jsonType: $value")
+  private def expected(what: String) = throw new ReadFailure(s"Expected $what but got $jsonType: ${reader.currentValue}")
 
-  private def matchOr[@specialized(Int, Double, Boolean) T: ClassTag](what: String): T = value match {
+  private def matchOr[@specialized(Boolean) T: ClassTag](what: String): T = reader.currentValue match {
     case t: T => t
     case _ => expected(what)
   }
 
-  def inputType: InputType = value match {
+  private def matchNumericString[T](toNumber: String => T): T = reader.currentValue match {
+    case ns: String if reader.isNumber || ns == "Infinity" || ns == "-Infinity" || ns == "NaN" =>
+      try toNumber(ns) catch {
+        case e: NumberFormatException => throw new ReadFailure(s"Invalid number format: $ns", e)
+      }
+    case _ =>
+      expected("numeric string")
+  }
+
+  def inputType: InputType = reader.currentValue match {
     case JsonStringInput.ListMarker => InputType.List
     case JsonStringInput.ObjectMarker => InputType.Object
     case null => InputType.Null
     case _ => InputType.Simple
   }
 
-  def readNull(): Null = if (value == null) null else expected("null")
-  def readString(): String = matchOr[String]("string")
-  def readBoolean(): Boolean = matchOr[Boolean]("boolean")
-  def readInt(): Int = matchOr[Int]("integer number")
-
-  def readLong(): Long = value match {
-    case i: Int => i
-    case s: String => s.toLong
-    case _ => expected("integer number or numeric string")
+  def readNull(): Null = if (reader.currentValue == null) null else expected("null")
+  def readString(): String = {
+    val expectedTpe = "string"
+    if (reader.isNumber) expected(expectedTpe)
+    matchOr[String](expectedTpe)
   }
-  def readDouble(): Double = matchOr[Double]("double number")
-
+  def readBoolean(): Boolean = matchOr[Boolean]("boolean")
+  def readInt(): Int = matchNumericString(_.toInt)
+  def readLong(): Long = matchNumericString(_.toLong)
+  def readDouble(): Double = matchNumericString(_.toDouble)
   def readBinary(): Array[Byte] = {
     val hex = matchOr[String]("hex string")
```

Comment:
This one should also check if `!isNumber`, I think we should have a dedicated method instead of using `matchOr[String]`